_posts/2017-04-04-what-every-developer-should-know-about-ir.md

```diff
@@ -0,0 +1,250 @@
+
+By the formula we can calculate that weight for "dog" is 0.75 `= 3 * (1 / 4)` for document 1 and 0.25 `= 1 * (1 / 4)` for document 2. From this we reason that document 1 is more relevant than documents 2 for term "dog" and thus should be raised higher in resultset. Note that weight is not necessarily used to reason if the document is found at all but to order the resultset.
+
+Matching algorithms is a vast subject, of which entire books are written, so it can't be thoroughly addressed here. Suggested further reading can be found at the end of this post.
+
+## Vector space vs. boolean model
+
+These models are usually kept clearly separate in research and literature. Most of the real life information retrieval systems are actually hybrid of these using ideas from various sources. For instance predominant Elasticsearch is a hybrid of boolean model and vector space model interpretation. It uses boolean model for matching documents, but utilizes vector space influenced sorting algorithm to sort result set. Postgres has a really good full text search support and it utilizes vector space model pretty cleanly.
+
+Google was originally really simple TF * IDF vector space model system with "page rank" added. Page rank added more weight to pages which are broadly linked from other pages and simirlarly diminished weight of the pages that were seldom linked. For now, we don't know basically anything how Google actually works.
+
+Other models of information retrieval include probabilistic model(s), latent semantic indexing, the binary independence retrieval, best-match retrieval and so on.
+
+## Terms, oh the terms
+
+*Terms* are the atoms of full text information retrieval, and they are called *terms* instead of *words* because usually all kinds of preprocessing is done before indexing and querying. *Terms* come out of *text* with the processes of *tokenisation* and *analysation*.
+
+When the boolean retrieval model was discussed, we already encountered problems with words as they inflect. In boolean model, we addressed to that with OR-facets or using wild cards, but that can be really hard to lazy users who have used to Google. With wild cards one can try to address the problem of word inflection.
+
+There must be something better, right?
+
+We mentioned *query expansion* already and next we take a look at *stop word lists* and *stemming*.
+
+### Stop word list
+
+Usually we just drop words that are not that meaningful in information retrieval context. Stop words in english usually consist of words such as: *a, an, the, none, some, he, she, it, his, her* and so on. When we tokenise raw text, the first thing to do is just drop these and do not add them to index at all. Similarly, when query is processed, the same exact stop word list should be used to filter query as well. It's nice, if user writes query `can I find me a little dog` which effectively queries for `find little dog`.
```

Comment:
_The_ stop word _list_ of _E_nglish 
When we tokenise _a_ raw text
add them to _the_ index
when _a_ query is processed
to filter _the_query as well